man/sort-dctrl.1
.TH SORT-DCTRL 1 2005-06-08 "Debian Project" "Debian user's manual"
\" Copyright (C) 2005 Antti-Juhani Kaijanaho <gaia@iki.fi>
\"      This program is free software; you can redistribute it and/or modify
\"      it under the terms of the GNU General Public License as published by
\"      the Free Software Foundation; either version 2 of the License, or
\"      (at your option) any later version.
\" 
\"      This program is distributed in the hope that it will be useful,
\"      but WITHOUT ANY WARRANTY; without even the implied warranty of
\"      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
\"      GNU General Public License for more details. 
\"  
\"      You should have received a copy of the GNU General Public License
\"      along with this program; see the file COPYING.  If not, write to
\"      the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
\"      Boston, MA 02111-1307, USA.
.SH NAME
sort\-dctrl \- sort Debian control files
.SH SYNOPSIS
.B sort\-dctrl
[options]
[
.IR file " ..."
]
.sp
.B sort\-dctrl
\-\-copying | \-\-help | \-\-version | \-ChV
.SH DESCRIPTION
The
.B sort\-dctrl
program sorts Debian control files according to specified criteria.
.PP
A Debian control (dctrl) file is a semistructured single-table
database stored in a machine-parseable text file.
.
Such a database
consists of a set of records; each record is a mapping from field
names to field content.
.
Textually, records are separated by empty
lines, while each field is encoded as one or more nonempty lines
inside a record.
.
A field starts with its name, followed by a colon,
followed by the field content.
.
The colon must reside on the first
line of the field, and the first line must start with no whitespace.
.
Subsequent lines, in contrast, always start with linear whitespace
(one or more space or tab characters).
.PP
The
.B sort\-dctrl
program recognizes two field types: string fields and version fields.
Version fields act also as numeric fields.  String fields are compared
according to strict lexicographical octet-by-octet comparison, after
ignoring any initial whitespace after the colon.  Version fields are
parsed and compared as Debian version numbers.  When comparing version
numbers, if a field content does not in fact conform to the version
number syntax, it compares less than any field content that does
conform, and equal to any other nonconforming field content.  The
.B sort\-dctrl
program assumes all fields are string fields unless told otherwise.
.PP
You can specify arbitrary number of keys for sorting, using the
.B \-k
option.  The keys are interpreted in a descending order of priority:
the first key specified is primary, the second key specified is
secondary, and so on.  If two records compare equal under the primary
key, then they are compared under the secondary key, and so on.  If no
keys are specified, a default key is assumed (the "Package" field with
no modifiers).
.SH OPTIONS
.IP "\fB\-k \fIkeyspec\fR, \fB\-\-key-spec=\fIkeyspec"
Specify one or more keys for sorting.  You may specify this option any
number of times.  The
.I keyspec
argument consists of a comma-separated
list of key specifications.  Each key specification consists of the
name of the field that serves as the key, optionally followed by a
colon and key modifiers.  The following key modifiers are supported:
.RS
.TP
.B r
Invert the comparison for this key, reversing the sorting order.
.TP
.B v
Treat this field as a version number field.
.TP
.B n
Treat this field as numeric, which currently is synonymous with 
.BR v .
.RE
.IP "\fB\-q\fR, \fB\-\-quiet\fR, \fB\-\-silent"
Output nothing to the standard output stream.  Instead, exit
immediately after finding the first match.
.IP "\fB\-l \fIlevel\fR, \fB\-\-errorlevel=\fIlevel"
Set log level to
.IR level .
.I level
is one of
.BR fatal ", " important ", " informational " and " debug ,
but the last may not be available,
depending on the compile-time options.  These categories are given
here in order; every message that is emitted when
.B fatal
is in effect, will be emitted in the
.B important
error level, and so on. The default is
.BR important .
.IP "\fB\-V\fR, \fB\-\-version"
Print out version information.
.IP "\fB\-C\fR, \fB\-\-copying"
Print out the copyright license.  This produces much output; be sure
to redirect or pipe it somewhere (such as your favourite pager).
.IP "\fB\-h\fR, \fB\-\-help"
Print out a help summary.
.SH EXAMPLES
Here are some sample invocations of the program:
.IP "sort\-dctrl /var/lib/dpkg/available"
Output the
.B dpkg
.I available
file sorted by the package name.
.IP "sort\-dctrl -k Version:v /var/lib/dpkg/available"
Output the
.B dpkg
.I available
file sorted in ascending order of
version numbers.
.IP "sort\-dctrl -k Version:vr /var/lib/dpkg/available"
Output the
.B dpkg
.I available
file sorted in descending order of
version numbers.
.IP "sort\-dctrl -k Package,Version:v /var/lib/dpkg/available"
Output the
.B dpkg
.I available
file sorted primarily in ascending order of
package names and secondarily in descending order of version numbers.
.IP "sort\-dctrl -k Installed-Size:n,Size:nr /var/lib/dpkg/available"
Output the
.B dpkg
.I available
file sorted primarily in ascending order of
installation sizes and secondarily in descending order of package
sizes.
.SH AUTHOR
The program and this manual page were written by Antti-Juhani
Kaijanaho <ajk@debian.org>.
.SH "SEE ALSO"
Debian Policy Manual.  Published as the Debian
package
.BR debian\-policy .
Also available in the Debian website.
.PP
.BR grep-dctrl (1)
\" Local variables:
\" mode: nroff
\" End:



NEWS
1.100 - a total rewrite
        new feature: boolean queries
        known regressions:
           + unacceptably slow search in full paragraph
           + no -Ffoo,bar,baz (equivalent boolean q available)
           + no internationalization

1.11 - added the -s option

1.10 - made -c synonymous with --count

1.7 - allowed several -F options and comma-separated list of
      fields as -F's argument

1.6 - disabled the -c option

1.3 - fixed a segfault related to the -c option
      deprecated the short -c option (use the long option instead)

1.2 - added the -v option (see manpage for details)
      New documentation file Compatibility.

1.0 - no news is good news


sort-dctrl/sort-dctrl.c
/*  dctrl-tools - Debian control file inspection tools
    Copyright © 2004, 2005, 2006, 2007, 2008, 2009, 2010 Antti-Juhani Kaijanaho

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <argp.h>
#include <locale.h>
#include <stdlib.h>
#include "misc.h"
#include "msg.h"
#include "i18n.h"
#include "ifile.h"
#include "para_bundle.h"
#include "sorter.h"

enum {
        OPT_SILENT = 256,
        OPT_MMAP
};

const char * argp_program_version = "sort-dctrl (dctrl-tools) " VERSION;
const char * argp_program_bug_address = MAINTAINER;

static struct argp_option options[] = {
	{ "copying",	       'C', 0,		    0, N_("Print out the copyright license.") },
	{ "errorlevel",	       'l', N_("LEVEL"),    0, N_("Set debugging level to LEVEL.") },
	{ "key-spec",          'k', N_("KEYSPEC"),  0, N_("Specify sort keys.") },
	{ "mmap",               OPT_MMAP, 0,        0, N_("Attempt mmapping input files") },
	{ 0 }
};

#define MAX_FNAMES 4096

struct arguments {
	/* Sort key specifications */
	keys_t keys;
	/* Number of file names seen.  */
	size_t num_fnames;
	/* File names seen on the command line.  */
	struct ifile fname[MAX_FNAMES];
};

static error_t parse_opt (int key, char * arg, struct argp_state * state)
{
	struct arguments * args = state->input;
	switch (key) {
	case 'C':
		if (!to_stdout (COPYING)) fail();
		exit(0);
	case 'k':
	{
		char *carg = strdup(arg);
		if (carg == 0) fatal_enomem(0);
		for (char *s = strtok(carg, ","); s != 0; s = strtok(0, ",")) {
			char *flags = strrchr(s, ':');
			if (flags == 0) {
				flags = "";
			} else {
				*(flags++) = '\0';
			}
			struct key key;
			key.field_inx = fieldtrie_insert(s)->inx;
			key.type = FT_STRING;
			key.reverse = false;
			for (char *p = flags; *p != '\0'; p++) {
				switch (*p) {
				case 'r':
					key.reverse = true;
					break;
				case 'v': case 'n':
					key.type = FT_VERSION;
					break;
				default:
					message(L_FATAL, 0, 
                                                _("invalid key flag"));
					fail();
				}
			}
                        debug("parse_opt: append key %d %d %d",
                              key.field_inx, key.type, key.reverse);
			keys_append(&args->keys, key);
		}
                debug_message("parse_opt: k", 0);
                break;
	}
	case 'l': 
	{
		int ll = str2loglevel(arg);
		if (ll < 0)
		{
			message(L_FATAL, 0, _("no such log level '%s'"), arg);
			fail();
		}
		set_loglevel(ll);
		debug_message("parse_opt: l", 0);
	}
		break;
	case OPT_MMAP:
		debug_message("parse_opt: mmap", 0);
		fsaf_mmap = 1;
		break;
	case ARGP_KEY_ARG:
		debug_message("parse_opt: argument", 0);
		{
			char const * s;
			if (args->num_fnames >= MAX_FNAMES) {
				message(L_FATAL, 0, _("too many file names"));
				fail();
			}
			s = strdup(arg);
			if (s == 0) fatal_enomem(0);
			args->fname[args->num_fnames++] =
				(struct ifile){ .mode = m_read, .s = s };
		}
		break;
	case ARGP_KEY_END:
	case ARGP_KEY_ARGS:  case ARGP_KEY_INIT: case  ARGP_KEY_SUCCESS:
	case ARGP_KEY_ERROR: case ARGP_KEY_FINI: case ARGP_KEY_NO_ARGS:
		debug_message("parse_opt: ignored", 0);
		break;
	default:
		return ARGP_ERR_UNKNOWN;
	}
	return 0;
}

static char progdoc [] = N_("sort-dctrl -- sort Debian control files");

static struct argp argp = { options, parse_opt, 0, progdoc };

int main(int argc, char * argv[])
{
	setlocale(LC_ALL, "");
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);

	fieldtrie_init();

	static struct arguments args;
	keys_init(&args.keys);
	msg_set_progname(argv[0]);
	argp_parse (&argp, argc, argv, ARGP_IN_ORDER, 0, &args);
        
        debug("number of keys: %zu", args.keys.nks);

	if (args.keys.nks == 0) {
		size_t inx = fieldtrie_insert("Package")->inx;
		keys_append(&args.keys, (struct key){ .field_inx = inx,
						      .type = FT_STRING,
						      .reverse = false });
	}

        if (args.num_fnames == 0) {
                args.fname[args.num_fnames++] = (struct ifile){
                        .mode = m_read,
                        .s = "-" 
                };
        }

	struct para_bundle pb;
	bundle_init(&pb);

	for (size_t i = 0; i < args.num_fnames; i++) {
		bundle_slurp(&pb, args.fname[i], false);
	}

	sort_bundle(&args.keys, &pb);
	size_t num_paras = bundle_size(&pb);
	para_t ** paras = bundle_vec(&pb);
	for (size_t i = 0; i < num_paras; i++) {
		struct fsaf_read_rv r = get_whole_para(paras[i]);
		fwrite(r.b, 1, r.len, stdout);
		putchar('\n');
		putchar('\n');
	}
	return 0;
}
